tests/tls_e2e/common/openssl_utility.cpp

```diff
@@ -0,0 +1,315 @@
+    der_len = (size_t)i2d_X509(crt, &buff);
+    if (der_len < 0)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "i2d_X509 failed(der_len=%d)\n", der_len);
+        goto done;
+    }
+
+    if (g_control_config.fail_oe_verify_attestation_certificate)
+        goto done;
+    // verify tls certificate
+    result = oe_verify_attestation_certificate(
+        der, der_len, enclave_identity_verifier, nullptr);
+
+    if (result != OE_OK)
+    {
+        OE_TRACE_ERROR(TLS_SERVER "result=%s\n", oe_result_str(result));
+        goto done;
+    }
+    ret = 1;
+done:
+
+    if (der)
+        free(der);
+
+    if (err != X509_V_ERR_DEPTH_ZERO_SELF_SIGNED_CERT)
+    {
+        OE_TRACE_ERROR(
+            TLS_SERVER "verifying SGX certificate extensions ... %s\n",
+            ret ? "succeeded" : "failed");
+    }
+    return ret;
+}
+
+int read_from_session_peer(
+    SSL*& ssl_session,
+    const char* payload,
+    size_t payload_length)
+{
+    int ret = -1;
+    unsigned char buf[200];
+    int bytes_read = 0;
+    do
+    {
+        int len = sizeof(buf) - 1;
+        memset(buf, 0, sizeof(buf));
+        bytes_read = SSL_read(ssl_session, buf, len);
+
+        if (bytes_read <= 0)
+        {
+            int error = SSL_get_error(ssl_session, bytes_read);
+            if (error == SSL_ERROR_WANT_READ)
+                continue;
+
+            OE_TRACE_ERROR("Failed! SSL_read returned error=%d\n", error);
+            ret = bytes_read;
+            break;
+        }
+
+        OE_TRACE_INFO(" %d bytes read from session peer\n", bytes_read);
+
+        // check to see if received payload is expected
+        if ((bytes_read != (int)payload_length) ||
+            (memcmp(payload, buf, (size_t)bytes_read) != 0))
+        {
+            OE_TRACE_ERROR(
+                "ERROR: expected reading %lu bytes but only "
+                "received %d bytes\n",
+                payload_length,
+                bytes_read);
+            ret = bytes_read;
+            goto exit;
+        }
+        else
+        {
+            OE_TRACE_INFO(
+                " received all the expected data from the session peer\n\n");
+            ret = 0;
+            break;
+        }
+
+        OE_TRACE_INFO(
+            "Verified: the contents of peer payload were expected\n\n");
+    } while (1);
+
+exit:
+    return ret;
+}
+
+int write_to_session_peer(
+    SSL*& ssl_session,
+    const char* payload,
+    size_t payload_length)
+{
+    int bytes_written = 0;
+    int ret = 0;
+
+    while ((bytes_written =
+                SSL_write(ssl_session, payload, (int)payload_length)) <= 0)
+    {
+        int error = SSL_get_error(ssl_session, bytes_written);
+        if (error == SSL_ERROR_WANT_WRITE)
+            continue;
+        OE_TRACE_INFO("Failed! SSL_write returned %d\n", error);
+        ret = bytes_written;
+        goto exit;
+    }
+
+    OE_TRACE_INFO("%lu bytes written to session peer \n\n", payload_length);
+exit:
+    return ret;
+}
+
+int load_oe_modules()
```

Comment:
This can be moved to utility.cpp and used by both mbedtls and openssl server&client.

Reply:
done